### ISSUE TITLE
Think about backup strategies

### ISSUE THREAD

**Issue by marvin-j97** [fjall-rs/fjall#52]
### Discussed in https://github.com/fjall-rs/fjall/discussions/50

<div type='discussions-op-text'>

<sup>Originally posted by **jeromegn** May 19, 2024</sup>
If one wanted to do a backup of the database, what's the best practice here? Is there a way to do an "online" (not shutting down the process using the database) backup?

Since there are many directories and files, I assume the safest way is to persist from the process, exit and then create a tarball of the whole directory.

With SQLite, for example, it's possible to run a `VACUUM INTO` that will create a space-efficient snapshot of the database. SQLite can support multiple processes though, so it's a different ball game.</div>

**Comment by xenacool** [comment]
Apache Paimon is an LSM format designed for writing to S3 compatible object storage while it's running. I really liked this author's articles on it.

https://jack-vanlightly.com/analyses/2024/7/3/understanding-apache-paimon-consistency-model-part-1

**Comment by marvin-j97** [comment]
> Has there been any more thought on this? I'm looking at using this crate but backups are necessary for my use case. I wouldn't mind trying out an implementation if theres been a general decision on the right approach.

The general idea is to use [RocksDB-like Checkpoints](https://github.com/facebook/rocksdb/wiki/Checkpoints) for **online** backups, I didn't want to add it in 1.x because I knew 2.x would add and change some stuff anyway. But to fully implement it, there will need to be some synchronization, because the journal(s) need to be copied, then all the disk segments (hard-linked if possible) and metadata files, *for all partitions*. Once the backup starts, no journal GC, journal writes, blob file rewrites, memtable flushes or compactions are allowed to complete until the backup is done, but reads should not be blocked. So I think there needs to be a keyspace-level RwLock...

For offline backups, simply using `cp -R` will work.

**Comment by marvin-j97** [comment]
I have added a todo list to the OP. I still have some stuff coming up for 2.4.0, so this isn't gonna make it in there, but I want to take a look at it in the near future... unless someone wants to take a look into it and contribute a possible solution, even if just a draft.

**Comment by marvin-j97** [comment]
@Svenskunganka I think it can be more granular. Copying the active journal should be quick, so we just need to lock it shortly. Then, we can already unlock the active journal - and then, we copy the sealed journals, they are just not allowed to be dropped by journal GC during all this time for consistency.

**Comment by marvin-j97** [comment]
V3 has had the necessary ground work to implement this in the somewhat near future because we can now take consistent snapshot across keyspaces even with compactions and writes going on.

**Comment by d11wtq** [comment]
> V3 has had the necessary ground work to implement this in the somewhat near future because we can now take consistent snapshot across keyspaces even with compactions and writes going on.

I implemented online backup using this mechanism myself:

1. Create a totally new database instance with the same configuration, in a temporary location.
2. Create a read snapshot of the current live database.
3. Iterate the entire keyspace ranges and insert into the temporary database.
5. Close/drop the temporary database.
6. Create an archive of the resulting directory.

That is a point-in-time clone of the original database and works for my use case.

**Comment by marvin-j97** [comment]
>That is a point-in-time clone of the original database and works for my use case.

Yes that may work for many scenarios, but the goal here is to be able to snapshot without having to read every data block, but instead just copying (or ideally hard-linking) the table + blob + metadata files. And then make that incremental, so future backups don't have to recopy everything every time.
However this is currently kind of blocked because while we can get a stable snapshot of the entire database, the incremental aspect is a bit more difficult. `lsm-tree` stores one version file *per tree/keyspace*, so I'm not yet sure how to atomically switch the version files. Otherwise an incremental backup could corrupt an existing backup.
